docs/adr/0025-flow-property-updates.md

```diff
@@ -0,0 +1,248 @@
+    * `description`
+      * Some TAMS applications may require descriptions to be set at creation time.
+    * `label`
+      * Some TAMS applications may require labels to be set at creation time.
+    * `updated_by`
+      * Set by TAMS using request metadata, e.g. HTTP headers.
+    * `tags`
+      * Some TAMS applications may require certain tags to be set at creation time.
+    * `read_only`
+    * `max_bit_rate`
+      * Some TAMS ingesters may calculate this over time and therefore it could be updated often.
+            Ideally this information is known at Flow creation so that Flow readers can decide on the required receiver buffer sizes for example.
+    * `avg_bit_rate`
+      * This could be calculated on-the-fly.
+            Ideally this information is known at Flow creation so that Flow readers know what data rate to expect.
+    * `flow_collection`
+      * The list of elements can be changed but not the elements themselves.
+* Properties that are set by TAMS and are read-only.
+  * Properties:
+    * `created`
+    * `metadata_updated`
+    * `segments_updated`
+    * `collected_by`
+    * `timerange`
+
+A Flow is deemed to be complete if all the properties that define the static characteristics for the Flow are set.
+This allows a TAMS API client to make decisions based on the information that is provided.
+
+A Flow could have incomplete or incorrect information.
+This means that the TAMS API should allow updates to properties that are normally expected to not change.
+The Flow metadata may be incomplete on creation but could be completed before the Flow is used.
+
+There could be systems that sit in front of the TAMS API that add restrictions to what can be updated.
+A system could for example require a client to have the "created_by permissions" for the Flow to enable updates to the `description` and `label` properties.
+
+## Considered Options
+
+* Option a1: Prevent updates of properties that should not change
+* Option a2: Continue to allow all properties to be updated with some exceptions
+
+* Option b1: Add endpoints for all properties that can be updated
+* Option b2: Add an endpoint with a path template matching the property that can be updated
+* Option b3: Continue to use the current Flow PUT and property updates
+
+* Option c1.1: Add a `/set` endpoint for PUTing to an arbitrary group of properties
+* Option c1.2: Also add a `/unset` endpoint for PUTing to remove an arbitrary group of properties
+* Option c2: Allow a PATCH update of a Flow
+* Option c3: Add endpoints for PUTing to defined groups of properties
+* Option c4: Continue to use the current Flow PUT and property updates
+
+## Decision Outcome
+
+The chosen options are: a2, b3 and c4.
+
+Option a2 is chosen to allow all properties (with some exceptions) to be corrected and not require a Flow to be complete on creation.
+It is expected that systems that sit in front of the TAMS API apply additional constraints on updates for a particular installation.
```

Comment:
There's a flexibility vs data integrity thing to think about I suppose: it would be frustrating if a client came to depend on the "adjust the Source ID after creation" mechanic only to run across an implementation that doesn't let you do that. Is there a compromise along the lines of "clients SHOULD NOT change these properties after creation, however implementations MAY permit it to allow operators to correct mistakes and simplify client implementations" or something?

Reply:
Makes sense. I've added similar words to the ADR.